frontends/concrete-python/concrete/fhe/compilation/module_compiler.py

```diff
@@ -234,6 +248,151 @@ def __call__(
+        return len(self.graph.input_nodes)
+
+    @property
+    def outputs_count(self) -> int:
+        """
+        Returns the number of outputs of the function.
+        """
+        assert self.graph
+        return len(self.graph.output_nodes)
+
+
+class NotComposable:
+    """
+    Composition policy that does not allow the forwarding of any output to any input.
+    """
+
+    def _get_rules_iter(self, _funcs: List[FunctionDef]) -> Iterable[CompositionRule]:
+        return []
+
+
+class ModuleComposable:
+    """
+    Composition policy that allows to forward any output of the module to any of its input.
+    """
+
+    def _get_rules_iter(self, funcs: List[Graph]) -> Iterable[CompositionRule]:
+        outputs = chain(
+            *[
+                map(CompositionClause.create, zip(repeat(f.name), range(len(f.output_nodes))))
+                for f in funcs
+            ]
+        )
+        inputs = chain(
+            *[
+                map(CompositionClause.create, zip(repeat(f.name), range(len(f.input_nodes))))
+                for f in funcs
+            ]
+        )
+        return map(CompositionRule.create, product(outputs, inputs))
+
+
+@runtime_checkable
+class WireOutput(Protocol):
+    """
+    A protocol for wire outputs.
+    """
+
+    def _get_outputs_iter(self) -> Iterable[CompositionClause]:
+        pass
+
+
+@runtime_checkable
+class WireInput(Protocol):
+    """
+    A protocol for wire inputs.
+    """
+
+    def _get_inputs_iter(self) -> Iterable[CompositionClause]:
+        pass
+
+
+class Output(NamedTuple):
+    """
+    The output of a given function of a module.
+    """
+
+    func: FunctionDef
+    pos: int
+
+    def _get_outputs_iter(self) -> Iterable[CompositionClause]:
+        return [CompositionClause(self.func.name, self.pos)]
+
+
+class AllOutputs(NamedTuple):
+    """
+    All the outputs of a given function of a module.
+    """
+
+    func: FunctionDef
+
+    def _get_outputs_iter(self) -> Iterable[CompositionClause]:
+        return map(
+            CompositionClause.create,
+            zip(repeat(self.func.name), range(self.func.outputs_count)),
+        )
+
+
+class Input(NamedTuple):
+    """
+    The input of a given function of a module.
+    """
+
+    func: FunctionDef
+    pos: int
+
+    def _get_inputs_iter(self) -> Iterable[CompositionClause]:
+        return [CompositionClause(self.func.name, self.pos)]
+
+
+class AllInputs(NamedTuple):
+    """
+    All the inputs of a given function of a module.
+    """
+
+    func: FunctionDef
+
+    def _get_inputs_iter(self) -> Iterable[CompositionClause]:
+        return map(
+            CompositionClause.create, zip(repeat(self.func.name), range(self.func.inputs_count))
+        )
+
+
+class Wire(NamedTuple):
+    """
+    A forwarding rule between an output and an input.
+    """
+
+    output: WireOutput
+    input: WireInput
+
+    # pylint: disable=protected-access
+    def _get_rules_iter(self, _) -> Iterable[CompositionRule]:
```

Comment:
Why make these private?

Reply:
Because it is not supposed to be used by users. How should I do it for you?

Reply:
It can be public IMO. Graph internals was not initially available, but now we provide an API for graph processors.